// @tanstack/react-form/dist/cjs/nextjs/createServerValidate.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const decodeFormdata = require("decode-formdata");
const error = require("./error.cjs");
const isFormValidationError = (error2) => {
  return typeof error2 === "object";
};
const createServerValidate = (defaultOpts) => async (formData, info) => {
  const { validatorAdapter, onServerValidate } = defaultOpts;
  const runValidator = async (propsValue) => {
    if (validatorAdapter && typeof onServerValidate !== "function") {
      return validatorAdapter().validateAsync(propsValue, onServerValidate);
    }
    return onServerValidate(propsValue);
  };
  const values = decodeFormdata.decode(formData, info);
  const onServerError = await runValidator({
    value: values,
    validationSource: "form"
  });
  if (!onServerError) return;
  const onServerErrorStr = onServerError && typeof onServerError !== "string" && isFormValidationError(onServerError) ? onServerError.form : onServerError;
  const formState = {
    errorMap: {
      onServer: onServerError
    },
    values,
    errors: onServerErrorStr ? [onServerErrorStr] : []
  };
  throw new error.ServerValidateError({
    formState
  });
};
const initialFormState = {
  errorMap: {
    onServer: void 0
  },
  values: void 0,
  errors: []
};
exports.createServerValidate = createServerValidate;
exports.initialFormState = initialFormState;
