// @tanstack/react-form/dist/esm/remix/createServerValidate.js
import { decode } from "decode-formdata";
import { ServerValidateError } from "./error.js";
const isFormValidationError = (error) => {
  return typeof error === "object";
};
const createServerValidate = (defaultOpts) => async (formData, info) => {
  const { validatorAdapter, onServerValidate } = defaultOpts;
  const runValidator = async (propsValue) => {
    if (validatorAdapter && typeof onServerValidate !== "function") {
      return validatorAdapter().validateAsync(propsValue, onServerValidate);
    }
    return onServerValidate(propsValue);
  };
  const values = decode(formData, info);
  const onServerError = await runValidator({
    value: values,
    validationSource: "form"
  });
  if (!onServerError) return;
  const onServerErrorStr = onServerError && typeof onServerError !== "string" && isFormValidationError(onServerError) ? onServerError.form : onServerError;
  const formState = {
    errorMap: {
      onServer: onServerError
    },
    values,
    errors: onServerErrorStr ? [onServerErrorStr] : []
  };
  throw new ServerValidateError({
    formState
  });
};
const initialFormState = {
  errorMap: {
    onServer: void 0
  },
  values: void 0,
  errors: []
};
export {
  createServerValidate,
  initialFormState
};
